Middleware jms

karelhala opened the pull request:
Add middleware jms to each server and make it possible to list all jms for all servers.

When clicking on refresh middleware provider jms are loaded successfully, however listing these jms is not yet working, because of error in `listicon_image` method.

![selection_048](https://cloud.githubusercontent.com/assets/3439771/17401529/758271f6-5a4e-11e6-876d-8f9d4a7433ae.png)
![selection_051](https://cloud.githubusercontent.com/assets/3439771/17481366/0c8409ee-5d7d-11e6-8082-a2d475d3e1a6.png)
![selection_052](https://cloud.githubusercontent.com/assets/3439771/17481365/0c747b96-5d7d-11e6-983a-c6908cdd2009.png)


- [x] Add tests
- [x] Add preview screenshots

karelhala commented:
@miq-bot add_label providers/hawkular, ui, wip

karelhala commented:
I was unable to find suitable icon for jms so I just rotated datasources icon and used is as placeholder, but we'll have to create specific icon for this.

@pilhuhn what do you think about fetching all recursive children when fetching server entities?

abonas commented:
> I was unable to find suitable icon for jms so I just rotated datasources icon and used is as placeholder, but we'll have to create specific icon for this.

@karelhala which icon is used for this in hawkular ui?

abonas commented on app/helpers/middleware_jms_helper/textual_summary.rb:
```diff
@@ -0,0 +1,46 @@
+    %i(tags)
+  end
+
+  def textual_name
```
I don't think this is needed, check include TextualMixins::TextualName part in other similar places, I think it already does that

abonas commented:
for the icon, you should use https://github.com/ManageIQ/manageiq/blob/master/app/assets/images/100/placeholder.png
(copy + rename it to the icon that needs design)
cc @epwinchell

abonas commented:
I'm not sure regarding the name Middleware_jms. Is it how it's called in hawkular?
as a start, jms should probably be all capitalized since it is an abbreviation.

abonas commented on app/models/manageiq/providers/hawkular/middleware_manager/refresh_parser.rb:
```diff
@@ -85,9 +85,10 @@ def uuid_provider_types
-          @ems.child_resources(eap.path).map do |child|
-            next unless child.type_path.end_with?('Deployment', 'Datasource')
+          @ems.child_resources(eap.path, true).map do |child|
+            next unless child.type_path.end_with?('Deployment', 'Datasource', 'JMS%20Topic', 'JMS%20Queue')
```
why is the need for `%20` - which component has the limitation that requires it?

karelhala replied:
In hawkular the path contains spaces, and they are encoded as %20 (JMS Topic -> JMS%20Topic), I thought about using `child.type_path.includes("JMS") && (child.type_path.includes("Topic") || child.type_path.includes("Queuq"))` but the rule with `ned_with` felt stronger and consistent. But fi you think it would be better to use contains, I can rewrite it, no problem.

abonas commented:
perhaps I missed it, but where's a differentiator/type in UI & backend for queue versus topic? 
For example, a user is looking at a jms summary page - how is it possible to know whether this is a queue or topic?

theute commented:
I was also wondering if we should separate queues and topics there. @pilhuhn @jdoyleoss ?

jdoyleoss commented:
Queues and Topics do have some different properties/metrics to them, so I think there is value in separating them. Eventually we will have similar but distinct representations, and it's better to start that way.

Fryguy commented:
What's a Jms?  Is this a generic term for all possible Middleware providers or is that a Hawkular specific thing (I generally prefer avoiding abbreviations unless it's an industry standard term).

jdoyleoss commented:
JMS is Java Message Service, it's one of the Java EE specifications and a subsystem in EAP servers. Within JMS there are queues and topics, which are the primary objects to model.

Fryguy commented:
Interesting...are all Middleware managers Java based?  If so, this is fine, but it struck me as something very specific to Java Middleware managers.

jdoyleoss commented:
The middleware provider (Hawkular) leverages EAP as it's runtime platform. So it's specific to EAP, and thus Hawkular.

karelhala commented:
> Interesting...are all Middleware managers Java based?

Not particular, it can be written in any language, but most of middlewares communicate trough messages.

Fryguy commented:
So, what I'm hearing is that an abstraction of this would be to call it middleware_message_service?

Fryguy commented on app/helpers/middleware_jms_helper/textual_summary.rb:
```diff
@@ -0,0 +1,46 @@
+
+  def textual_jndi_name
+    {:label => _('JNDI Name'),
+     :value => @record.properties['JNDI Name']}
```
On my comment about generalization, this also leads to other hawkular specifics like here, which is weird if another middleware provider is created.

pilhuhn commented:
Actually JMS is the Java version of a Message Oriented Middleware(MOM). Those also come in other flavors (e.g. MQTT for IoT), but the notion of Topics and Queues are the same for all MOMs. In the context of Java/JBoss JMS is certainly the term people want to use.

abonas commented:
> Queues and Topics do have some different properties/metrics to them, so I think there is value in separating them. Eventually we will have similar but distinct representations, and it's better to start that way.

Please clarify what kind of separation - completely separate to different entities? Have a single representation with different 'type' field? (I'd go with the latter)

abonas commented:
> So, what I'm hearing is that an abstraction of this would be to call it middleware_message_service?

I propose middleware_message_destination

pilhuhn commented:
Yes, *destination* is the 'superclass' for Topics and Queues.

karelhala commented:
I've fixed problems with decorator and things proposed by @abonas (thanks for pointing those flaws out).

There is still work needed on test suites and we should add more information to message detail.

Once we will agree on naming this center I'll rename it. I forgot that we will support more than just Java middleware so I went blindly for JMS, but this should be named better.

Do we agree on renaming it from `Middleware jms` to `Middleware message destination`?

abonas commented:
> Once we will agree on naming this center I'll rename it. I forgot that we will support more than just Java middleware so I went blindly for JMS, but this should be named better.
> 
> Do we agree on renaming it from Middleware jms to Middleware message destination?

@jdoyleoss please clarify what is the direction here so the work on this PR can proceed. thanks.
is there any place where jms should be mentioned at all as well as topic/queue? (imo it can be in the type field "jms queue" and "jms topic" for Java users clarity)

pilhuhn commented:
I think we should go for a general "superclass" of "messaging destination", which can then be sub-divided into message queue and message topic. Those can be on the MiQ-level like the generic middleware_server. This way we can support any message-oriented-middleware in the future.

Then inside of Hawkular we can subclass this for specifics of the EAP7' Artemis Queues and Topics. Here at that level we can name them JMS Queue and JMS Topic if we want.

karelhala commented:
The problem here is that JMS have ques and topics, but Python might have different name conventions. So I propose this (item highlights are broken, because I just copy pasted html on console, so provider is highlighted instead of Messaging):
![selection_039](https://cloud.githubusercontent.com/assets/3439771/17023471/d1678f16-4f53-11e6-943d-d97ff16282d6.png)

Either this, or as @pilhuhn proposed (Egain, provider is highlighted instead of Messaging):
![selection_040](https://cloud.githubusercontent.com/assets/3439771/17023759/3bfcbc7e-4f55-11e6-91bb-b6771669e7b9.png)

First solution will allow for combining same middleware's server messages, the other will combine message types together.

I'll let you choose @pilhuhn @abonas @jdoyleoss which is better.

abonas commented:
My concern with "messaging" is that if the entities are named 'messages', that sounds like they are the actual 'units' (similar to events) that are sent, versus queues and topics which are the channels/destination - those that allow them to be sent.
@jdoyleoss the terminology here is your call.

pilhuhn commented:
Messaging is ok to me. The thing that does the work is usually called 'Message oriented Middleware' (or MoM). A message is what is sent. See e.g. https://en.wikipedia.org/wiki/Message-oriented_middleware (or https://en.wikipedia.org/wiki/Message_oriented_middleware, but the first one is better).

We could rename Topic to Fan-out or Pub/Sub. RabbitMQ seems to use the term fanout (and the example is in Pyhton https://www.rabbitmq.com/tutorials/tutorial-three-python.html )
Queue is ubiquitous.

abonas commented:
> Messaging is ok to me. The thing that does the work is usually called 'Message oriented Middleware' (or MoM). A message is what is sent. See e.g. https://en.wikipedia.org/wiki/Message-oriented_middleware (or https://en.wikipedia.org/wiki/Message_oriented_middleware, but the first one is better).
> 
> We could rename Topic to Fan-out or Pub/Sub. RabbitMQ seems to use the term fanout (and the example is in Pyhton https://www.rabbitmq.com/tutorials/tutorial-three-python.html )
> MQTT calls them Topics too. 
> Queue is ubiquitous.
> 
> So I guess for people in the MoM space, Topic and Queue are known terms.

the question is how many people, especially existing hawkular users and those who come from java world will understand the move from topic to fanout

jdoyleoss commented:
I think Messaging is the right top level term to use, and that we should use Topics and Queues under that. I think these are both the correct Java terms, generally understood, and even applicable outside of Java EE.

abonas commented:
> I think Messaging is the right top level term to use, and that we should use Topics and Queues under that. I think these are both the correct Java terms, generally understood, and even applicable outside of Java EE.

messaging it is.
do you want messaging, topic, queue prefixed with middleware like the rest of the entities?

jdoyleoss commented:
Yes, let's keep it consistent.  Thanks

martinpovolny commented on config/initializers/inflections.rb:
```diff
@@ -31,6 +31,7 @@
   inflect.singular(/Quota$/, "Quota")
   inflect.plural(/quota$/, "quotas")
   inflect.irregular("container_quota", "container_quotas")
+  inflect.irregular("MiddlewareJms", "MiddlewareJms")
```
@mzazrivec : do we want to use this file?

karelhala replied:
This has been moved to `config/application.rb` however since we swithced from middleware jms to middleware messaging we no longer need to add inflections in this PR. (plural of messaging is [messagings](https://en.wiktionary.org/wiki/messaging))

abonas commented:
@karelhala please rebase this and also the icon/images needs to be replaced by the proper placeholder image

abonas commented on app/helpers/application_helper/toolbar/middleware_messaging_center.rb:
```diff
@@ -0,0 +1,36 @@
+    select(
+      :middleware_messaging_monitoring_choice,
+      'product product-monitoring fa-lg',
+      t = N_('Monitoring'),
```
why is this relevant - do we have metrics collected for this at the moment?

karelhala replied:
Well from what I understand we collect metrics for all resources, and JMS is just different type of resource. But if it's not needed I can remove this.

abonas replied:
it is correct that there is a mechanism to collect metrics. But at the same time, there should be certain configurations made to specify which metrics should be collected, which is not part of this PR.

abonas commented on app/models/middleware_messaging.rb:
```diff
@@ -0,0 +1,12 @@
+  acts_as_miq_taggable
+  serialize :properties
+
+  include LiveMetricsMixin
```
shouldn't this be done in a PR that adds metrics? cc @lucasponce

lucasponce replied:
To simplify the PR, yes I guess it can be easier to add support for metrics in a following step.

abonas commented on app/models/middleware_messaging.rb:
```diff
@@ -0,0 +1,12 @@
+
+  include LiveMetricsMixin
+
+  def metrics_capture
```
same comment here as above regarding metrics

abonas commented on app/views/configuration/_ui_2.html.haml:
```diff
@@ -59,14 +59,21 @@
                :middlewaredeployment                 => ["middleware_deployment_show_list", _('Middleware Deployments')],
-               :middlewaredatasource                 => ["middleware_datasource_show_list", _('Middleware Datasources')]}.each do |resource, (feature, view_name)|
+               :middlewaredatasource                 => ["middleware_datasource_show_list", _('Middleware Datasources')],
+               :middlewaremessaging                => ["middleware_messaging_show_list", _('Middleware Messaging')]}}.each do |resource, (feature, view_name)|
```
please align the `=>` , I think rubocop would complain about this

abonas commented on app/views/middleware_messaging/edit.html.haml:
```diff
@@ -0,0 +1 @@
+= render :partial => 'shared/views/ems_common/form'
```
edit seems not relevant

abonas commented on app/views/middleware_messaging/new.html.haml:
```diff
@@ -0,0 +1 @@
+= render :partial => 'shared/views/ems_common/form'
```
New action also seems irrelevant - we are not creating jms queues and topics from miq at the moment (and not as part of this PR)

abonas commented on config/locales/en.yml:
```diff
@@ -788,6 +788,7 @@ en:
       ManageIQ::Providers::Azure::NetworkManager::NetworkPort:              Network Port (Microsoft Azure)
       ManageIQ::Providers::Openstack::NetworkManager::NetworkPort:          Network Port (OpenStack)
       MiddlewareDatasource:     Middleware Datasource
+      MiddlewareMessaging:            Middleware Messaging
```
please align this properly

abonas commented on db/fixtures/miq_product_features.yml:
```diff
@@ -3522,6 +3522,56 @@
+      :description: Display Individual Middleware Messaging
+      :feature_type: view
+      :identifier: middleware_messaging_show
+    - :name: Utilization
```
in line with my other metrics related comments - this PR afaik doesn't introduce the metrics support as well, so this is less relevant

abonas commented on product/views/MiddlewareMessaging.yaml:
```diff
@@ -0,0 +1,74 @@
+#
+
+# Report title
+title: Middleware Messagings
```
perhaps Messaging would be more correct English wise..

abonas commented on app/helpers/middleware_messaging_helper/textual_summary.rb:
```diff
@@ -0,0 +1,23 @@
+  #
+
+  def textual_group_properties
+    %i(name nativeid)
```
what about the type? how can user know if this is queue or topic?

karelhala replied:
Good question, I tried fetching type from hawkular inventory resource, but I am unable to access `type` value in `refresh_parser.rb`. I looked in [entities.rb](https://github.com/hawkular/hawkular-client-ruby/blob/56c16c41f1074cc7599e5c3be81f76f1291d13c2/lib/hawkular/inventory/entities.rb#L61) and it looks like there's missing `attr_reader :type`, hence no public access to this property.

However I can do `to_h` and read the type variable. So what should I do? New PR in hawkular-ruby-client` and make type variable publicly accessible (make it part of the API) or just use `to_h` and access type this way?

karelhala commented:
@abonas Removed all utilization settings and added messaging type to this PR. Could you please look if I missed anything? Thank you.

abonas commented:
@karelhala the tests are failing, please check it. thanks

karelhala commented:
Ok the test gave me a lot of headache.

The problem was with refresh tests and VCR settings. What I did in this PR is that we are no longer calling children, but recursiveChildren. I tried update VCR using recording new episodes, but that was not good, since feedId changes (feedId in VCR was different thant the one on my computer).

I had to delete refresher.yml and run recording this VCR egain. I bumped into problem with not having middleware domains so I followed this guide (with @Jiri-Kremser help): https://github.com/ManageIQ/manageiq/pull/10048

So if anyone will change something in a way that it will call new URL he'll have to do the same steps.

abonas commented:
> I had to delete refresher.yml and run recording this VCR egain. I bumped into problem with not having middleware domains so I followed this guide (with @Jiri-Kremser help): #10048
> 
> So if anyone will change something in a way that it will call new URL he'll have to do the same steps.

why was there a problem with not having mw domains in regards to jms?

you mentioned some solution/steps above taken from the #10048, worth to document this with more detail, this is not very clear how to solve it from reading the comment. (at least to me :smiley: )

also, since #10148, I believe some code can be erased from controllers due to adding reusable shared code. please check it. thanks

karelhala commented:
> why was there a problem with not having mw domains in regards to jms?

I had to record brand new VCR and these tests checks for mw domains and domain groups, so I had to reproduce same steps as in #10048 it's not required to use same steps when running this, just to clarify what to look at when refresher tests are failing.
The problem was that when VCR tests were recorded they used different feedId, and now when fetching resources recursive method is used so new calls were made but with bad feedId, so I removed whole VCR file, added mw domains and server groups as described in #10048 egain just to record new VCR with recursive resources fetch.

I'll check #10148 and see if I can simplify this PR.

abonas commented:
@miq-bot assign @rubenvp8510 
@rubenvp8510 please review this PR, thanks!

rubenvp8510 commented:
I tested it and works well, this looks good.

abonas commented:
@karelhala there are a lot of commits in this PR. Some of them should not be here, such as "Fix tests" or "Fix styling" - because they are relevant internally in the PR, but meaningless once it's merged to master. The reason is that when someone will look at git history, seeing "fix tests" or "fix styling" will sound like the master branch needed fixing, while it was not the case here. So, it might be hard to fix it now in this PR, but looking forward please avoid such commits. Using 'git --amend' is very useful in cases like this.

abonas commented:
I also validated this.
Few comments and questions. @jdoyleoss 
1. is Messagings (in plural) a valid name? It doesn't sound like proper English to me, but I might be wrong on this.
2. It looks like the type (JMS queue/topic) is displayed always twice - in the name and in the type column. Do we want this duplicity? Is it *always* part of the name?
![jms](https://cloud.githubusercontent.com/assets/6277245/17649527/40c42e60-6240-11e6-9dba-f6244ac199a3.png)

abonas commented on app/helpers/application_helper/toolbar/middleware_messaging_center.rb:
```diff
@@ -0,0 +1,19 @@
+                     button(
+                       :middleware_messaging_tag,
+                       'pficon pficon-edit fa-lg',
+                       N_('Edit Tags for this Middleware messaging'),
```
@jdoyleoss @serenamarie125 I have a general question regarding such strings - should Middleware be with capital letter or not? Messaging - same question - capital or not? (this will of course apply to align in all other entities as well)

abonas commented on app/models/manageiq/providers/hawkular/middleware_manager/refresh_parser.rb:
```diff
@@ -195,6 +207,20 @@ def parse_deployment(server, deployment)
+      def parse_messaging(server, messaging, config)
+        data = {
+          :name              => messaging.name,
+          :middleware_server => server,
```
it is better to leverage here the `parse_base_item `method, like done here:
https://github.com/ManageIQ/manageiq/blob/master/app/models/manageiq/providers/hawkular/middleware_manager/refresh_parser.rb#L195

abonas commented on app/models/manageiq/providers/hawkular/middleware_manager/refresh_parser.rb:
```diff
@@ -195,6 +207,20 @@ def parse_deployment(server, deployment)
+          :messaging_type    => messaging.to_h['type']['name']
+        }
+        if !config.empty? && !config['value'].empty? && config['value'].respond_to?(:except)
+          data[:properties] = config['value'].except('Username', 'Password')
```
is there username/password in all jms configurations? (just a general question)

abonas commented on app/models/middleware_messaging.rb:
```diff
@@ -0,0 +1,6 @@
+class MiddlewareMessaging < ApplicationRecord
+  belongs_to :ext_management_system, :foreign_key => "ems_id"
+  belongs_to :middleware_server, :foreign_key => "server_id"
```
is this also correct in domain mode? (that jms topic/queue belongs to only one server) @Jiri-Kremser @pilhuhn

pilhuhn replied:
I think in the Domain mode this may not be true, as a destination is set up on the level of a profile, which is associated with the server group and thus applies to all servers in that server group. 

I think reporting it for each server in a server group may be ok, as effectively they would share the same destination and either use it for load balancing or such.

abonas replied:
> I think reporting it for each server in a server group may be ok, as effectively they would share the same destination and either use it for load balancing or such.

will it duplicate it in miq if reported more than once? (from several servers)

karelhala replied:
It would be duplicated when showing all messaging. However do we want to filter these out? What if some other middleware behaves differently? And in domain mode the destination of JMS would not be same for each server?

abonas replied:
we need to reflect hawkular inventory, so if this is reported as duplicated, we should show as duplicated imo. @pilhuhn what shall you say?

abonas commented on config/routes.rb:
```diff
@@ -1191,6 +1191,43 @@
+        show
+        show_list
+        tl_chooser
+        update
```
I assume it's not relevant for jms at the moment

abonas commented on db/fixtures/miq_product_features.yml:
```diff
@@ -3543,6 +3543,44 @@
+    :feature_type: admin
+    :identifier: middleware_messaging_admin
+    :children:
+    - :name: Remove
```
afaik we don't support removing for jms at the moment. @pilhuhn ?

pilhuhn replied:
Certainly not the subsystem. And I don't even know about individual destinations. As we don't add them, I think we als don't delete them - wdyt?

abonas commented on db/schema.yml:
```diff
@@ -4541,6 +4541,17 @@ middleware_domains:
 - ems_id
 - created_at
 - updated_at
+middleware_messagings:
```
I don't think you're supposed to commit schema.yml, it's an auto generated file. @jrafanie ?

jrafanie replied:
I see others have committed to this file.  I think that's ok. @carbonin can you confirm?

carbonin replied:
Correct, `schema.yml` needs to be committed. `schema.rb` should not be committed.

pilhuhn commented on app/helpers/ems_middleware_helper/textual_summary.rb:
```diff
@@ -11,7 +11,7 @@ def textual_group_properties
     # Order of items should be from parent to child
     items = []
-    items.concat(%i(middleware_servers middleware_deployments middleware_datasources))
+    items.concat(%i(middleware_servers middleware_deployments middleware_datasources middleware_messagings))
```
Shouldn't this follow the other textual_summary ones and just be 
`%i(middleware_servers... middleware_messagins)` without the `items` array?
I know that this like this for a long time.

abonas replied:
I think you are correct. `%i(...)` already creates an array of symbols, should be sufficient. 
(and yes, it was not added in this PR, so it can be certainly fixed in a separate followup)

pilhuhn commented on app/presenters/menu/default_menu.rb:
```diff
@@ -141,6 +141,13 @@ def middleware_menu_section
+            'middleware_messaging',
+            {:feature => 'middleware_messaging_show_list'},
+            '/middleware_messaging'
+          ),
```
While I think the long lines in that file are / could be an issue, please adjust to the style of the file for consistency

abonas commented:
@karelhala fyi regarding #10461 - less code to write...

karelhala commented:
@abonas 
> 1. is Messagings (in plural) a valid name? It doesn't sound like proper English to me, but I might be wrong on this.

It felt strange for me as well, but it looks like [messaging's](https://en.wiktionary.org/wiki/messaging) plural is [messagings](https://en.wiktionary.org/wiki/messagings#English)

> 2. It looks like the type (JMS queue/topic) is displayed always twice - in the name and in the type column. Do we want this duplicity? Is it always part of the name?

This is because name of topic/queue in wildfly is `JMS queue [TypeOfMessaging]` and type is `JMS Queue/Topic` at least this is the information I receive from inventory. We could parse the type out of the name if we want thou.

abonas commented:
Edit tags operation yields several errors
![exceptiontags](https://cloud.githubusercontent.com/assets/6277245/17700647/612b9d4c-63cf-11e6-805f-8b0ec323bf1a.png)

abonas commented on app/models/manageiq/providers/hawkular/middleware_manager.rb:
```diff
@@ -131,9 +133,9 @@ def server_groups(feed, _domain)
     end
 
-    def child_resources(resource_path)
+    def child_resources(resource_path, recursive = false)
```
why is the recursive parameter needed? (what is difference from other entities)

karelhala replied:
It's because of hierarchy in inventory, JMS are kept in recursive children, where at first level (non recursive) is only messaging server (this is how it's actually named in inventory, it's just resource with child entities), and queues / topics are located under this entity.

abonas commented:
> @abonas
> 
> is Messagings (in plural) a valid name? It doesn't sound like proper English to me, but I might be wrong on this.
> It felt strange for me as well, but it looks like messaging's plural is messagings

ok lets keep it :smiley: 
> 
> It looks like the type (JMS queue/topic) is displayed always twice - in the name and in the type column. Do we want this duplicity? Is it always part of the name?
> This is because name of topic/queue in wildfly is JMS queue [TypeOfMessaging] and type is JMS Queue/Topic at least this is the information I receive from inventory. We could parse the type out of the name if we want thou.

for now leave it as is, we can always do it later

abonas commented:
tags work as expected now. there's some schema related failure, please have a look

abonas commented:
ok, all tests pass, good! @rubenvp8510 do you have any comments on the PR or it can be moved to miq?

abonas commented:
@karelhala please remove the wip signs.

karelhala commented:
@miq-bot add_label wip

karelhala commented:
@miq-bot remove_label wip

abonas commented:
@miq-bot assign @chessbyte

abonas commented:
@karelhala please rebase/fix conflicts as needed. thank you

karelhala commented:
@abonas rebasing was quite a challenge since domains were changing similar files.

abonas commented:
@karelhala I'm getting this error when refreshing after applying this PR: "unknown attribute 'feed' for ManageIQ::Providers::Hawkular::MiddlewareManager::MiddlewareMessaging. "
please check

karelhala commented:
That's because I had to add feed to migration script. I've added it to creation script, but thinking about it that might be a bad idea, because when someone already did tried this he'd have to `rake db:rollback STEP=someNumber` and `rake db:update`.

What do you think? Might be better to create new db update which will add feed to db, or keeping it this way?

abonas commented:
> That's because I had to add feed to migration script. I've added it to creation script, but thinking about it that might be a bad idea, because when someone already did tried this he'd have to rake db:rollback STEP=someNumber and rake db:update.
> 
> What do you think? Might be better to create new db update which will add feed to db, or keeping it this way?

since this wasn't merged yet, I'd go with putting them in the same migration script introduced in this PR so it will be merged as "one piece". I'll rerun the necessary scripts and retry.

abonas commented on app/views/configuration/_ui_2.html.haml:
```diff
@@ -68,7 +68,8 @@
                :middlewaredomain                     => ["middleware_domain_show_list",        _('Middleware Domains')],
-               :middlewareservergroup                => ["middleware_server_groups_show_list", _('Middleware Server Groups')]}.each do |resource, (feature, view_name)|
+               :middlewareservergroup                => ["middleware_server_groups_show_list", _('Middleware Server Groups')],
+               :middlewarejms                        => ["middleware_messaging_show_list",     _('Middleware Messaging')]}.each do |resource, (feature, view_name)|
```
the key should be middlewaremessaging, not middlewarejms, probably missed in renaming?

abonas commented:
LGTM except for [one comment](https://github.com/ManageIQ/manageiq/pull/9514#discussion_r76254947).

abonas commented:
@Fryguy please review, or, if you are overloaded, @chessbyte please assign someone for review/merge? it contains schema changes and we'd like to complete with those asap. thank you

martinpovolny commented:
The UI coode looks good. But this needs to be tested in the UI.

@abonas : is this tested? or can you give me access to a provider where I can test this? @karelhala? @Jiri-Kremser ?

abonas commented:
> The UI coode looks good. But this needs to be tested in the UI.
> 
> @abonas : is this tested? or can you give me access to a provider where I can test this? @karelhala? @Jiri-Kremser ?

I tested it in UI [last week](https://github.com/ManageIQ/manageiq/pull/9514#issuecomment-242411324)
